src/lib/components/dateSelector/index.tsx

```diff
@@ -59,229 +24,233 @@ export interface DateSelectorProps {
-      date.day !== undefined
+      // is not day input
+      index > 0 &&
+      // has clicked backspace or arrow left
+      ['Backspace', "ArrowLeft"].includes(event.key) && 
+      // is focused at the first character of the input
+      inputSelectionStart === 0 && inputSelectionEnd === 0
     ) {
-      if (
-        calendarDateValue === undefined ||
-        date.day !== calendarDateValue.day ||
-        date.month !== calendarDateValue.month ||
-        date.year !== calendarDateValue.year
+      const prevInput = itemsRef.current?.[index - 1];
+
+      event.preventDefault();
+      prevInput?.focus();
+      prevInput?.setSelectionRange(0, 3);
+    }
+
+    if (
+        // is not year input
+        index < 2
+        // has clicked arrow right
+        && event.key === "ArrowRight"
+        // is focused at the last character of the input value
+        && inputSelectionStart === currentInput.value.length
       ) {
-        onChange(
-          calendarDateToISODate({
-            day: date.day,
-            month: date.month,
-            year: date.year,
-          })
-        );
-      }
+      const nextInput = itemsRef.current?.[index + 1];
+      
+      event.preventDefault();
+      nextInput?.focus();
+      nextInput?.setSelectionRange(0, index === 1 ? 4 : 3);
     }
-  }, [date]); //eslint-disable-line react-hooks/exhaustive-deps
+  }
+
+  const handleOnKeyUp = (event: KeyboardEvent<HTMLInputElement>, index: number) => {
+    const currentInput = itemsRef.current?.[index];
+    const inputSelectionStart = currentInput?.selectionStart;
 
-  const handleOnChange = (key: keyof CalendarDate, v: number) => {
-    const newValue = { ...date, [key]: v };
     if (
-      key !== 'day' &&
-      newValue.month !== undefined &&
-      newValue.year !== undefined &&
-      newValue.day !== undefined
-    ) {
-      const cappedDays = Math.min(
-        daysInMonthOfYear({ month: newValue.month, year: newValue.year }),
-        newValue.day
-      );
-      setDate({ ...newValue, day: cappedDays });
-    } else {
-      setDate(newValue);
+        // is not year input
+        index < 2
+        // is a number key
+        && /^\d+$/.test(event.key)
+        && (
+          // is focused at the last character of the input value
+          inputSelectionStart === currentInput.maxLength 
+          // or month value is over 1 or day value is over 3
+          || Number(currentInput.value) > (index === 1 ? 1 : 3)
+        )
+      ) {
+      const nextInput = itemsRef.current?.[index + 1];
+      
+      event.preventDefault();
+      event.stopPropagation();
+      
+      nextInput?.focus();
+      nextInput?.setSelectionRange(0, index === 1 ? 4 : 3);
     }
+  }
 
-    setOpenCalendar(false);
-  };
+  const getInputProps = (key: keyof CalendarDate, index: number) => ({
+    'data-cy': `date-selector-${key}`,
+    'data-testid': `date-selector-${key}`,
+    className: styles[`${key}Input`],
+    id: key,
+    name: key,
```

Comment:
How can we use this as uncontrolled component?

To give you more context, on the task engine we are adding a specific name and id to each input.
For example, a text input would have an id like this: `input-f8556724-a703-455e-a6f5-16b0d16bb26d`

But for the date input, we want the full date, so we are doing a tiny hack [by using a hidden field with the full date](https://github.com/getPopsure/app/blob/e2069cacaed963aaeb99f0afa3a122f0a7cf7b6d/src/features/taskEngine/components/CustomTask/inputs/DateInput.tsx#L17).

I'm wondering if we should have such hidden field already inside the component instead of the individual name for each input.

Happy to discuss this a bit further. We don't need to change anything for this PR.